mmengine/model/wrappers/test_time_aug.py

```diff
@@ -0,0 +1,145 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from abc import abstractmethod
+from typing import Dict, List, Union
+
+import torch
+import torch.nn as nn
+
+from mmengine import MODELS
+from mmengine.structures import BaseDataElement
+from ..base_model import BaseDataPreprocessor
+
+# multi-batch inputs processed by different augmentations from the same batch.
+EnhancedBatchInputs = List[Union[torch.Tensor, List[torch.Tensor]]]
+# multi-batch data samples processed by different augmentations from the same
+# batch. The outer list stands for different augs and the inner list stands for
+# batch.
+EnhancedBatchDataSamples = List[List[BaseDataElement]]
+DATA_BATCH = Union[Dict[str, Union[EnhancedBatchInputs,
+                                   EnhancedBatchDataSamples]], tuple, dict]
+MergedDataSamples = List[BaseDataElement]
+
+
+@MODELS.register_module()
+class BaseTTAModel:
+    """Base model for inference with test-time augmentation.
+
+    ``BaseTTAModel`` is a wrapper for inference given multi-batch data.
+    It implements the :meth:`forward` for multi-batch data inference.
+    `multi-batch` data means data processed by different augmentation
+    from the same batch.
+
+    All subclasses should implement :meth:`merge_preds` for results fusion.
+
+    During test time augmentation, the data processed by
+    :obj:`mmcv.transforms.TestTimeAug`, and then collated by
+    `pseudo_collate` will have the following format:
+
+    .. code-block::
+
+        result = dict(
+            inputs=[
+                [image1_aug1, image2_aug1],
+                [image1_aug2, image2_aug2]
+            ],
+            data_samples=[
+                [data_sample1_aug1, data_sample2_aug1],
+                [data_sample1_aug2, data_sample2_aug2],
+            ]
+        )
+
+    ``image{i}_aug{j}`` means the i-th image of the batch, which is
+    augmented by the j-th augmentation.
+
+    ``BaseTTAModel`` will collate the data to:
+
+     .. code-block::
+
+        data1 = dict(
+            inputs=[image1_aug1, image2_aug1],
+            data_samples=[data_sample1_aug1, data_sample2_aug1]
+        )
+
+        data2 = dict(
+            inputs=[image1_aug2, image2_aug2],
+            data_samples=[data_sample1_aug2, data_sample2_aug2]
+        )
+
+    ``data1`` and ``data2`` will be passed to model, and the results will be
+    merged by :meth:`merge_preds`.
+
+    Note:
+        :meth:`merge_results` is an abstract method, all subclasses should
+        implement it.
+
+    Args:
+        module (BaseModel): Tested model.
+        data_preprocessor (BaseDataPreprocessor or dict, optional): The
+            pre-process config for :class:`BaseDataPreprocessor`.
+    """
+
+    def __init__(self,
+                 module: nn.Module,
+                 data_preprocessor: Union[dict, BaseDataPreprocessor,
+                                          None] = None):
+
+        super().__init__()
+        if isinstance(module, nn.Module):
```

Comment:
Should we check whether the module is BaseModel rather than `nn.Module` because `test_step` of `BaseTTAModel` calls the `test_step` method of `BaseModel`? Besides, suggest renaming `module` to `model`.